MySQL Geometry throw unhandled exception when buffer is null

sushantdhiman opened the pull request:
Fix #4953

mickhansen commented on test/integration/data-types.test.js:
```diff
@@ -274,3 +274,30 @@ describe(Support.getTestDialectTeaser('DataTypes'), function() {
   });
 });
+
+describe(Support.getTestDialectTeaser('DataTypes') + " MySQL", function() {
```
`Support.getTestDialectTeaser` already appends mysql.
Would it be possible to simply have a general empty coordinates test for all dialects?

sushantdhiman replied:
I am running this test for all dialects which supports `GEOMETRY`. Will wait for CI to finish to see if something goes wrong on other supported databases.

mickhansen commented on test/integration/data-types.test.js:
```diff
@@ -274,3 +274,30 @@ describe(Support.getTestDialectTeaser('DataTypes'), function() {
   });
 });
+
+describe(Support.getTestDialectTeaser('DataTypes'), function() {
```
Isn't this describe the same as the one on the top of the file? Might aswell just remove it.

sushantdhiman commented:
Hi @mickhansen , Its done. Let me know if there is anything else to improve.

mickhansen commented on test/integration/data-types.test.js:
```diff
@@ -273,4 +273,29 @@ describe(Support.getTestDialectTeaser('DataTypes'), function() {
+        expect(users[0]).to.be.not.null;
+      });
+   }
+ });
```
This test needs to be expanded a bit.
Change the description to `should parse an empty GEOMETRY field`.

Add a comment at `field: { type: "Point", coordinates: [] }` about this generating a `NULL` insert for MySQL.
Add a verification on the fetched user that values are as expected.

sushantdhiman commented:
Hi @mickhansen , Changed tests and added changelog entry. I have rebased with latest `3.15` release as well. Can you please restart the CI, its not starting due to some reason.

mickhansen commented:
CI is queued atm it looks like

sushantdhiman commented:
Hmm... there are no other pull requests in queue. Others are finished. Both of my jobs blocked. Lets wait and see if they eventually timeouts.

sushantdhiman commented:
Hi @mickhansen , What should I return in case of null coordinates ?

`Postgres` returns ```{  type: 'Point',  coordinates: null }```
`MySQL` returns `null`

mickhansen commented:
Good question, might want @janmeier's opinion here.

janmeier commented:
I think always returning an object makes the most sense - that still encodes the type information, (Point / Polygon etc), even if the value is null

sushantdhiman commented:
Sorry tried to fixed it by editing files via Github , Didn't worked. I will add this change tomorrow.

sushantdhiman commented:
I have been working on this and I realized the `null` buffer returned by `node-mysql` is the issue behind it. `node-mysql` should return the valid `wkb` type buffer which contain information about GEOMETRY data even though its empty.

I have added an issue here https://github.com/felixge/node-mysql/issues/1310, @mickhansen If you want you can mark it blocking due to `node-mysql`. I will work on `node-mysql` for a solution.

sushantdhiman commented:
MySQL lacks support for `POINT EMPTY` as per this https://dev.mysql.com/worklog/task/?id=2381, This means any invalid / empty data is stored as `NULL` in the database. 

There is no way to know if that was `Point` or `LineString` or any other type of geometry data. Thus until MySQL implements `POINT EMPTY` we can't return `{ type: 'Point', coordinates: [] }`. 

If you want we can fix current issue by returning the `NULL` for now. When MySQL implements `POINT EMPTY` we can return the desired output `{ type: 'Point', coordinates: [] }`

cc @mickhansen @janmeier

mickhansen commented:
It should probably be null for mysql then, agree @janmeier?

sushantdhiman commented:
@janmeier ping

janmeier commented:
Yes

sushantdhiman commented:
Please restart the CI, two pull request are not started from last 30 min @janmeier

janmeier commented:
I haven't found the switch for restarting travis yet ;) - Just give it some time

sushantdhiman commented:
@janmeier All green, Let me know if PR need other changes.

janmeier commented on lib/dialects/mysql/data-types.js:
```diff
@@ -60,8 +60,15 @@ module.exports = function (BaseTypes) {
+    value = value.buffer();
+
+    //MySQL doesn't support POINT EMPTY, https://dev.mysql.com/worklog/task/?id=2381
+    if(value === null){
```
Missing space here

(Really wish this could be handled by the linter @mickhansen , I've noticed you commenting on it more and more :)). JSHint can't do it, dunno if other linters can

mickhansen replied:
Hmm, i think eslint does it, but that's for ES/Babel code

janmeier replied:
But since es5 is a subset of es2015, that shouldn't be a problem :)

janmeier commented on test/integration/data-types.test.js:
```diff
@@ -273,4 +273,35 @@ describe(Support.getTestDialectTeaser('DataTypes'), function() {
+        //This case throw unhandled exception
+        return User.findAll();
+      }).then(function(users){
+        if (dialect === 'mysql' || dialect === 'mariadb') {
```
`Support.dialectIsMySQL`

sushantdhiman replied:
will `Support.dialectIsMySQL` work for `mariadb` as well?

janmeier replied:
Yes

janmeier commented:
LGTM, just some tiny comments

mickhansen commented:
Just restarted the builds that failed. No reason to keep closing/opening the PR :)

sushantdhiman commented:
There is no restart button. Where I can find it ?

mickhansen commented:
Only me and Jan have access to it.

sushantdhiman commented:
Ok, Do you think we still need `io js v1` based tests ?  CI take too much time :disappointed:

mickhansen commented:
At the moment yes. When we publish 4.0 i'd like to move to only test against Node v4+

mickhansen commented:
A bigger problem is builds failing because of intermittent test errors.

sushantdhiman commented:
Yes this happens to me like everyday. Sometime CI not able to login to DB, sometime a query left running.

mickhansen commented:
Green now, will let @janmeier merge

sushantdhiman commented:
Thanks

janmeier commented:
Thanks @sushantdhiman :)

sushantdhiman commented:
Thanks @janmeier and @mickhansen
